### PR TITLE
CKKSTensor ops

### PR DESCRIPTION
**Pull request by bcebere**
## Description
 - [x] Support sum by any axis.
 - [x] In case of batching, supports sum only by the batching axis. 
 - [x] Fix batching for other dimensions than 0.
 - [x] Add negate operation.
 - [x] Add square operation.
 - [x] gtest+ python tests

fixes https://github.com/OpenMined/TenSEAL/issues/165
fixes https://github.com/OpenMined/TenSEAL/issues/162
fixes https://github.com/OpenMined/TenSEAL/issues/172

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: tenseal/cpp/tensors/ckkstensor.cpp [hunk 1]
```diff
@@ -188,7 +193,50 @@ shared_ptr<CKKSTensor> CKKSTensor::mul_plain_inplace(const double& to_mul) {
...
+
+            size_t new_idx = 0;
+            for (size_t pidx = 0; pidx < pos.size(); ++pidx)
+                new_idx += new_strides[pidx] * pos[pidx];
+
+            tenseal_context()->evaluator->add_inplace(new_data[new_idx],
+                                                      _data[idx]);
+        }
+        _data = new_data;
+    }
```
Comment (youben11) [inline tenseal/cpp/tensors/ckkstensor.cpp:237]: I think there is an add_many function in SEAL, that takes a vector of ciphertexts. Can it be helpful here? We wouldn't then need to initialize ciphertexts to 0.

File: tenseal/cpp/tensors/ckkstensor.cpp [hunk 2]
```diff
@@ -188,7 +193,50 @@ shared_ptr<CKKSTensor> CKKSTensor::mul_plain_inplace(const double& to_mul) {
...
 
 shared_ptr<CKKSTensor> CKKSTensor::sum_inplace(size_t axis) {
-    // TODO
+    if (axis >= _shape.size()) throw invalid_argument("invalid axis");
+
+    auto new_shape = _shape;
+    new_shape.erase(new_shape.begin() + axis);
+
+    if (_batch_size) {
+        if (axis != 0) throw invalid_argument("unsupported argument");
```
Comment (youben11) [inline tenseal/cpp/tensors/ckkstensor.cpp:202]: Why isn't it possible to sum other axis when there is a batch_axis? It should be still possible to do so without affecting the batch_size, keeping it the same while summing up over the other axis

File: tenseal/tensors/__init__.py [hunk 3]
```diff
@@ -120,7 +122,7 @@ def ckks_vector_from(context, data):
     )
 
 
-def ckks_tensor(context, tensor, scale=None):
+def ckks_tensor(context, tensor, scale=None, batch=True):
```
Comment (youben11) [inline tenseal/tensors/__init__.py:125]: I think beginner users might not expect the batching to happen, they might find it weird somehow. Can we make the default to false here?

File: tenseal/binding.cpp [hunk 4]
```diff
@@ -27,6 +28,8 @@ void bind_plain_tensor(py::module &m, const std::string &name) {
         .def("shape", &type::shape)
         .def("strides", &type::strides)
         .def("size", &type::size)
+        .def("batch", &type::batch)
+        .def("__len__", &type::size)
```
Comment (youben11) [inline tenseal/binding.cpp:32]: Most tensor libraries seems to return the shape[0] in len(tensor). Should we implement this the same way?

File: tenseal/binding.cpp [hunk 5]
```diff
@@ -27,6 +28,8 @@ void bind_plain_tensor(py::module &m, const std::string &name) {
         .def("shape", &type::shape)
         .def("strides", &type::strides)
         .def("size", &type::size)
+        .def("batch", &type::batch)
```
Comment (youben11) [inline tenseal/binding.cpp:31]: Can we please have some tests for this feature?

**Review by youben11 — CHANGES_REQUESTED** [review]
This looks amazing!

File: tenseal/binding.cpp (hunk 4 above)
Comment (bcebere) [reply tenseal/binding.cpp:32]: Isn't this already implemented? or I'm missing something

https://github.com/OpenMined/TenSEAL/blob/master/tenseal/cpp/tensors/plain_tensor.h#L233

File: tenseal/binding.cpp (hunk 5 above)
Comment (bcebere) [reply tenseal/binding.cpp:31]: There are already c++ tests https://github.com/OpenMined/TenSEAL/blob/ckks_tensor_improvements/tests/cpp/tensors/plaintensor_test.cpp#L124

I will add python tests as well.

File: tenseal/binding.cpp (hunk 4 above)
Comment (youben11) [reply tenseal/binding.cpp:32]: Oh yeah, my bad.

**Review by youben11 — APPROVED** [review]
Amazing! Thanks Bogdan

**Pull request merged**